Driver docs update (#1334)
docker/buildx: docs/guides/drivers/docker-container.md

# Docker container driver

The buildx Docker container driver allows creation of a managed and customizable
BuildKit environment in a dedicated Docker container.

Using the Docker container driver has a couple of advantages over the default
Docker driver. For example:

- Specify custom BuildKit versions to use.
- Build multi-arch images

> - Advanced options for [cache import and export](../cache/index.md)

Reviewer: ```suggestion
- Advanced options for [cache import and export](index.md)
```

Author: I wanted to link to the cache docs overview. Bad idea? :)

Reviewer: > I wanted to link to the cache docs overview.

Ah I thought your were linking to the same folder.

> Bad idea? :)

Yes atm we can't link to pages outside the current folder for remote pages. You have to use absolute links:

```suggestion
- Advanced options for [cache import and export](https://docs.docker.com/build/building/cache/)
```

- Build multi-arch images, see [QEMU](#qemu)
- Advanced options for
  [cache import and export](https://docs.docker.com/build/building/cache/)

## Synopsis

Run the following command to create a new builder, named `container`, that uses
the Docker container driver:

```console
$ docker buildx create --name container --driver docker-container
container
```

The following table describes the available driver-specific options that you can
pass to `--driver-opt`:

| Parameter       | Value  | Default          | Description                                                                                |
| --------------- | ------ | ---------------- | ------------------------------------------------------------------------------------------ |
| `image`         | string |                  | Sets the image to use for running BuildKit.                                                |
| `network`       | string |                  | Sets the network mode for running the BuildKit container.                                  |
| `cgroup-parent` | string | `/docker/buildx` | Sets the cgroup parent of the BuildKit container if Docker is using the `cgroupfs` driver. |

## Usage

> When you run a build, Buildx pulls the specified `moby/buildkit` image from

Reviewer 2: ```suggestion
When you run a build, Buildx pulls the specified `image` (by default, [`moby/buildkit`](https://hub.docker.com/r/moby/buildkit))
```
Oops my original writing wasn't super clear here :eyes:

...
```

## Loading to local image store

Unlike when using the default `docker` driver, images built with the
`docker-container` driver must be explicitly loaded into the local image store.
Use the `--load` flag:

```bash

> $ docker buildx build . --load -t <image> --builder=container

Reviewer 3: Why?

Author: oops yeah I was not getting syntax highlighting in my editor with `console` so I used bash while editing, I will change it back...

Author: Or maybe you meant why I changed the position of the build context arg :-) 

only to align with the cache docs where many examples are multiline and the dot at the end becomes awkward and sometimes missed

Reviewer: Having context set there is not common and does not reflect the usage:

```
Usage:  docker buildx build [OPTIONS] PATH | URL | -
```

This is what we use atm across all our documentation and examples. Would prefer to keep it as it is.

```

The image becomes available in the image store when the build finishes:

```console
$ docker image ls
REPOSITORY                       TAG               IMAGE ID       CREATED             SIZE
<image>                          latest            adf3eec768a1   2 minutes ago       197MB
```

### QEMU

The `docker-container` driver supports using [QEMU](https://www.qemu.org/) (user
mode) to build non-native platforms. Use the `--platform` flag to specify which
architectures that you want to build for.

For example, to build a Linux image for `amd64` and `arm64`:

```console
$ docker buildx build . \
  --builder=container \
  --platform=linux/amd64,linux/arm64 \
  -t <registry>/<image> \
  --push
```

> **Warning**
>
> QEMU performs full-system emulation of non-native platforms, which is much
> slower than native builds. Compute-heavy tasks like compilation and
> compression/decompression will likely take a large performance hit.

## Further reading

For more information on the Docker container driver, see the
[buildx reference](https://docs.docker.com/engine/reference/commandline/buildx_create/#driver).

If want to explore builders running on a remote server, see the

> [Kubernetes driver](./kubernetes.md) and the [Remote driver](./remote.md).

Reviewer 2: Ahhh this isn't exactly true :see_no_evil: `docker-container` "can" do things on a remote server by using a different docker context. I can't find where we document this explicitly anywhere, but we should explain more of the interaction between `buildx` and `docker context` in the `docker.md` page in the future, and we could cross-link to there from here.
